File: csharp/complex-numbers/ComplexNumbers.cs
public readonly struct ComplexNumber
{
    private readonly double _real;
    private readonly double _imaginary;
    public ComplexNumber(double real, double imaginary)
    {
        _real = real;
        _imaginary = imaginary;
    }
    //This is for allow passing plain number (int/double) where ComplexNumber is expected
    public static implicit operator ComplexNumber(double value)
    => new ComplexNumber(value, 0);
    public double Real()
    => _real;

    public double Imaginary()
    => _imaginary;

    public ComplexNumber Add(ComplexNumber other)
    => new ComplexNumber(_real + other._real, _imaginary + other._imaginary);

    public ComplexNumber Sub(ComplexNumber other)
    => new ComplexNumber(_real - other._real, _imaginary - other._imaginary);

    public ComplexNumber Mul(ComplexNumber other)
    {
        double a = _real;
        double b = _imaginary;
        double c = other._real;
        double d = other._imaginary;
        return new ComplexNumber(a * c - b * d, a * d + b * c);
    }

    public ComplexNumber Div(ComplexNumber other)
    {
        double a = _real;
        double b = _imaginary;
        double c = other._real;
        double d = other._imaginary;
        double denom = c * c + d * d;
        return new ComplexNumber((a * c + b * d) / denom, (b * c - a * d) / denom);
    }

    public double Abs()
    => Math.Sqrt(Math.Pow(_real, 2) + Math.Pow(_imaginary, 2));

    public ComplexNumber Conjugate()
    => new ComplexNumber(_real, -_imaginary);
    
    public ComplexNumber Exp()
    {
        //Formula: e^(a + bi) = e^a * (cos(b) + i * sin(b))
        double expReal = Math.Exp(_real);
        return new ComplexNumber(expReal * Math.Cos(_imaginary), expReal * Math.Sin(_imaginary));
    }
}

File: javascript/complex-numbers/complex-numbers.js
//
// This is only a SKELETON file for the 'Complex Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ComplexNumber {
  constructor(real, imag) {
    this._real = real;
    this._imag = imag;
  }

  get real() {
    return this._real;
  }

  get imag() {
    return this._imag
  }

  add(other) {
    return new ComplexNumber(
      this.real + other.real, 
      this.imag + other.imag
    );
  }

  sub(other) {
    return new ComplexNumber(
      this.real - other.real,
      this.imag - other.imag
    );
  }

  mul(other) {
    const o = other instanceof ComplexNumber ? other : new ComplexNumber(other, 0);
    const realPart = this.real * o.real - this.imag * o.imag;
    const imagPart = this.imag * o.real + this.real * o.imag;
    return new ComplexNumber(
      realPart === 0 ? 0 : realPart, 
      imagPart === 0 ? 0 : imagPart
    );
  }

  div(other) {
    const o = other instanceof ComplexNumber ? other : new ComplexNumber(other, 0);
    const denominator = o.real ** 2 + o.imag ** 2;
    let realPart = (this.real * o.real + this.imag * o.imag) / denominator;
    let imagPart = (this.imag * o.real - this.real * o.imag) / denominator;
    return new ComplexNumber(
      realPart === 0 ? 0 : realPart, 
      imagPart === 0 ? 0 : imagPart
    );
  }

  get abs() {
    return Math.sqrt(this.real ** 2 + this.imag ** 2);
  }

  get conj() {
    return new ComplexNumber(this.real, this.imag === 0 ? 0 : -this.imag);
  }

  get exp() {
    const expReal = Math.exp(this.real);
    return new ComplexNumber(
      expReal * Math.cos(this.imag),
      expReal * Math.sin(this.imag)
    );
  }
}
